osf_tests/test_draft_registration.py

```diff
@@ -0,0 +1,719 @@
+                'comments': [],
+                'value': '',
+                'extra': []
+            }
+        }
+        draft.update_metadata(new_metadata)
+        draft.save()
+        # To preserve both workflows, if update_metadata is called,
+        # a flattened version of that metadata is stored in
+        # registration_responses
+        assert draft.registration_responses == {
+            'looked': 'Yes',
+            'datacompletion': 'No, data collection has not begun',
+            'comments': ''
+        }
+
+    def test_update_metadata_tracks_changes(self, project):
+        draft = factories.DraftRegistrationFactory(branched_from=project)
+
+        draft.registration_metadata = {
+            'foo': {
+                'value': 'bar',
+            },
+            'a': {
+                'value': 1,
+            },
+            'b': {
+                'value': True
+            },
+        }
+        changes = draft.update_metadata({
+            'foo': {
+                'value': 'foobar',
+            },
+            'a': {
+                'value': 1,
+            },
+            'b': {
+                'value': True,
+            },
+            'c': {
+                'value': 2,
+            },
+        })
+        draft.save()
+        for key in ['foo', 'c']:
+            assert key in changes
+
+    def test_update_registration_responses(self, project):
+        schema = RegistrationSchema.objects.get(
+            name='OSF-Standard Pre-Data Collection Registration',
+            schema_version=2
+        )
+        draft = factories.DraftRegistrationFactory(registration_schema=schema, branched_from=project)
+        registration_responses = {
+            'looked': 'Yes',
+            'datacompletion': 'No, data collection has not begun',
+            'comments': ''
+        }
+        draft.update_registration_responses(registration_responses)
+        draft.save()
+        # To preserve both workflows, if update_metadata is called,
+        # a flattened version of that metadata is stored in
+        # registration_responses
+        assert draft.registration_metadata == {
+            'looked': {
+                'comments': [],
+                'value': 'Yes',
+                'extra': []
+            },
+            'datacompletion': {
+                'comments': [],
+                'value': 'No, data collection has not begun',
+                'extra': []
+            },
+            'comments': {
+                'comments': [],
+                'value': '',
+                'extra': []
+            }
+        }
+
+    def test_has_active_draft_registrations(self):
+        project, project2 = factories.ProjectFactory(), factories.ProjectFactory()
+        factories.DraftRegistrationFactory(branched_from=project)
+        assert project.has_active_draft_registrations is True
+        assert project2.has_active_draft_registrations is False
+
+    def test_draft_registrations_active(self):
```

Comment:
Should this also test some conditions under which a draft_registration wouldn't appear in this list?

Reply:
That was the idea behind the finished draft but I realized that this test shouldn't have been passing and there were some unfounded assumptions about what properties factories will assign to the created model instance. Updated it to accurately check to make sure non-active draft registrations don't show up.